### PR TITLE
fix: w3c trace header test

### PR DESCRIPTION
**Pull request by dmehala**
## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

File: tests/parametric/test_headers_tracecontext.py [hunk 1]
```diff
@@ -774,22 +773,21 @@ def test_tracestate_w3c_p_extract(self, test_agent, test_library):
         assert case1["meta"]["_dd.parent_id"] == "0123456789abcdef"
 
         assert case2["name"] == "p_invalid"
-        assert case2["meta"]["_dd.parent_id"] == "XX!X"
+        assert "_dd.parent_id" not in case2["meta"]
```
Comment (mabdinur) [inline tests/parametric/test_headers_tracecontext.py:776]:
```suggestion
        assert case2["meta"]["_dd.parent_id"] == "0000000000000000"
```

Good catch here. Invalid parent_ids should not be propagated. `_dd.parent_id` should be set to `0000000000000000`.

File: tests/parametric/test_headers_tracecontext.py [hunk 2]
```diff
@@ -774,22 +773,21 @@ def test_tracestate_w3c_p_extract(self, test_agent, test_library):
...
 
         assert case3["name"] == "datadog_headers_used_in_propagation"
         assert case3["trace_id"] == 5
         assert case3["parent_id"] == 11
         assert "_dd.parent_id" not in case3["meta"]
 
         assert case4["name"] == "p_not_propagated"
-        assert case4["meta"]["_dd.parent_id"] == "0000000000000000"
+        assert "_dd.parent_id" not in case4["meta"]
```
Comment (mabdinur) [inline tests/parametric/test_headers_tracecontext.py:784]: According to the extraction spec `_dd.parent_id` tag should always be set. If p is not propagated or is invalid.
```suggestion
        assert case4["meta"]["_dd.parent_id"] == "0000000000000000"
```

**Review by mabdinur — COMMENTED** [review]
I will update the ruby, python, and node tracer to better align with the spec. Can we move case2 and case4 to their own system tests and then skip it for all failing languages. This way we can maintain test coverage for case1 and case3 which should be passing for all languages. 

Thanks for catching this

File: tests/parametric/test_headers_tracecontext.py (hunk 1 above)
Comment (dmehala) [reply tests/parametric/test_headers_tracecontext.py:776]: I might had overlooked the RFC. If `p` is invalid, why `_dd.parent_id` should be set to "0000000000000000"?

Comment (bouwkast) [reply tests/parametric/test_headers_tracecontext.py:776]: Validation of the propagated `p` sub-key on extraction wasn't considered to be important - when it gets to the backend it'll need to be validated there regardless of whether we validated it or not as we don't have exclusive control over the contents of `tracestate`.

But fine with me either way, if validation whenever it gets extracted makes more sense then 👍

Comment (dmehala) [reply tests/parametric/test_headers_tracecontext.py]: @bouwkast not sure to understand. In scenario 2, the RFC spec that `_dd.parent_id` should be `0000000000000000`. Do you mean the backend will still have to validate `p`, thus `_dd.parent-id=0000000000000000` or `_dd.parent_id` is equivalent? If yes, then I think not sending `_dd.parent_id` is better resource wise.

Comment (dmehala) [reply tests/parametric/test_headers_tracecontext.py]: Thank you @mabdinur.

I think the confusion comes from:

> The RFC also states that when an **_invalid_** `p` value is extracted `_dd.parent_id` should be set to `0000000000000000`.

To know if `p` is invalid, we must run some validation process on `p`. 

My understanding of that _validation_ was:
> - The p sub-key MUST be injected into the dd list-member of the tracestate.
> - The value of p MUST be a 16-character lowercase hexadecimal encoded string.
> - The value of p MUST be set to the ID of the currently active span.
> - The value of p MUST be a non-zero value (0000000000000000 is invalid).
> - A : (colon) character MUST be placed between the key and the value. For example, p:0123456789abcdef

While the rest of the team understanding was (I guess):
> - The initial value MUST be a hexadecimal string 16 characters in length.

In the end, that's a misinterpretation on me. The initial intent is to centralize the validation, I think it's a good thing idea but at the expense of sending unnecessary data in some cases. 

The goal is not to add more work for everybody, I hereby declare this PR as unnecessary. Thank you for your patience ;)

Comment (mabdinur) [reply tests/parametric/test_headers_tracecontext.py]: Oh I am sorry for not clarifying this earlier. I will close this PR 😄

**Comment by mabdinur** [comment]
This change is no longer required. 

Context: https://github.com/DataDog/system-tests/pull/2407#discussion_r1591915860

File: tests/parametric/test_headers_tracecontext.py (hunk 1 above)
Comment (dmehala) [reply tests/parametric/test_headers_tracecontext.py]: Actually, this change is still required as `p:XX!X` is less than 16, thus not valid. right? 👀 haha I'm sorry.

**Pull request closed without merging**